File: testsuite/tests/ada/0005-arrays/cpp/main.cpp
#include <iostream>

#include "test.h"

using gnatpolyglot::ada::arrays::polyglot_array;

void test_integer_arrays() {
    std::cout << "-- integer arrays --\n";

    // A function returning an unconstrained array, owned on the C++ side.
    polyglot_array<int32_t> arr = test::f_u_1();
    std::cout << "bounds: " << arr.get_begin() << " " << arr.get_end() << "\n";
    std::cout << "content: [" << arr.get(1) << " " << arr.get(2) << " "
              << arr.get(3) << "]\n";
    std::cout << "sum: " << test::f_u_2(arr) << "\n";

    // Write elements back, then read them through a parameter.
    arr.set(1, 4);
    arr.set(2, 5);
    arr.set(3, 6);
    std::cout << "content: [" << arr.get(1) << " " << arr.get(2) << " "
              << arr.get(3) << "]\n";
    std::cout << "sum: " << test::f_u_2(arr) << "\n";

    // An `in out` parameter: Ada mutates the shared buffer in place.
    test::out_proc(arr);
    std::cout << "doubled:";
    for (auto &x : arr) {
        std::cout << " " << x;
    }
    std::cout << "\n";

    // An array allocated and filled on the C++ side (through the iterator).
    polyglot_array<int> owned(1, 5);
    int v = 0;
    for (auto &x : owned) {
        x = (v += 2);
    }
    std::cout << "owned:";
    for (auto &x : owned) {
        std::cout << " " << x;
    }
    std::cout << "\n";
}

void test_integer_widths() {
    std::cout << "-- integer widths --\n";

    polyglot_array<int8_t> bytes = test::make_bytes();
    std::cout << "bytes:";
    for (int i = bytes.get_begin(); i <= bytes.get_end(); i++) {
        std::cout << " " << (int) bytes.get(i);
    }
    std::cout << "\n";

    polyglot_array<int16_t> shorts = test::make_shorts();
    std::cout << "shorts:";
    for (int i = shorts.get_begin(); i <= shorts.get_end(); i++) {
        std::cout << " " << shorts.get(i);
    }
    std::cout << "\n";

    // The 64-bit values exceed 32 bits, proving the elements are not truncated.
    polyglot_array<int64_t> longs = test::make_longs();
    std::cout << "longs:";
    for (int i = longs.get_begin(); i <= longs.get_end(); i++) {
        std::cout << " " << longs.get(i);
    }
    std::cout << "\n";
}

void test_floating_point_arrays() {
    std::cout << "-- floating-point arrays --\n";

    polyglot_array<float> floats = test::make_floats();
    std::cout << "floats:";
    for (auto &x : floats) {
        std::cout << " " << x;
    }
    std::cout << "\n";
    std::cout << "sum: " << test::sum_floats(floats) << "\n";

    // An `in out` parameter mutates the shared buffer in place.
    test::scale_floats(floats);
    std::cout << "scaled:";
    for (auto &x : floats) {
        std::cout << " " << x;
    }
    std::cout << "\n";
    std::cout << "sum: " << test::sum_floats(floats) << "\n";

    // A float array allocated on the C++ side.
    polyglot_array<float> owned(1, 3);
    owned.set(1, 0.25f);
    owned.set(2, 0.5f);
    owned.set(3, 0.75f);
    std::cout << "owned:";
    for (auto &x : owned) {
        std::cout << " " << x;
    }
    std::cout << "\n";

    polyglot_array<double> doubles = test::make_doubles();
    std::cout << "doubles:";
    for (auto &x : doubles) {
        std::cout << " " << x;
    }
    std::cout << "\n";
    std::cout << "sum: " << test::sum_doubles(doubles) << "\n";
}

void test_boolean_arrays() {
    std::cout << std::boolalpha << "-- boolean arrays --\n";

    polyglot_array<bool> bools = test::make_bools();
    std::cout << "bools:";
    for (auto &x : bools) {
        std::cout << " " << x;
    }
    std::cout << "\n";
    std::cout << "any: " << test::any_true(bools) << "\n";

    // An `in out` parameter flips each element in place.
    test::negate_bools(bools);
    std::cout << "negated:";
    for (auto &x : bools) {
        std::cout << " " << x;
    }
    std::cout << "\n";
    std::cout << "any: " << test::any_true(bools) << "\n";

    // A boolean array allocated on the C++ side.
    polyglot_array<bool> owned(1, 2);
    owned.set(1, true);
    owned.set(2, false);
    std::cout << "owned:";
    for (auto &x : owned) {
        std::cout << " " << x;
    }
    std::cout << "\n";
}

void test_record_arrays() {
    std::cout << "-- record arrays --\n";

    polyglot_array<test::MyInt> arr = test::my_int_arr_func();
    std::cout << "bounds: " << arr.get_begin() << " " << arr.get_end() << "\n";
    std::cout << "content:";
    for (auto &el : arr) {
        std::cout << " " << el.get_i();
    }
    std::cout << "\n";

    // An `in out` parameter mutates the records in place.
    test::my_int_arr_proc(arr);
    std::cout << "tripled:";
    for (auto &el : arr) {
        std::cout << " " << el.get_i();
    }
    std::cout << "\n";
    test::print_image(arr);

    // A record array allocated and filled on the C++ side.
    polyglot_array<test::MyInt> owned(1, 3);
    int v = 0;
    for (auto &el : owned) {
        el = test::MyInt((v += 1) * 10);
    }
    std::cout << "owned:";
    for (auto &el : owned) {
        std::cout << " " << el.get_i();
    }
    std::cout << "\n";
}

int main() {
    test_integer_arrays();
    test_integer_widths();
    test_floating_point_arrays();
    test_boolean_arrays();
    test_record_arrays();
}

File: testsuite/tests/ada/0005-arrays/java/Main.java
import com.adacore.libtest.test.TestPackage;
import com.adacore.libtest.test.MyInt;
import com.adacore.gnatpolyglot.runtime.ada2java.*;

public class Main {

    static void testIntegerArrays() {
        System.out.println("-- integer arrays --");

        // A function returning an unconstrained array.
        IntegerArray arr = TestPackage.fU1();
        System.out.println("bounds: " + arr.getBegin() + " " + arr.getEnd());
        System.out.println("content: [" + arr.getUnslided(1) + " "
            + arr.getUnslided(2) + " " + arr.getUnslided(3) + "]");
        System.out.println("sum: " + TestPackage.fU2(arr));

        // Write elements back, then read them through a parameter.
        arr.setUnslided(1, 4);
        arr.setUnslided(2, 5);
        arr.setUnslided(3, 6);
        System.out.println("content: [" + arr.getUnslided(1) + " "
            + arr.getUnslided(2) + " " + arr.getUnslided(3) + "]");
        System.out.println("sum: " + TestPackage.fU2(arr));

        // An `in out` parameter: Ada mutates the shared buffer in place.
        TestPackage.outProc(arr);
        System.out.print("doubled:");
        for (var x : arr) {
            System.out.print(" " + x);
        }
        System.out.println();

        // An array allocated and filled on the Java side.
        IntegerArray owned = new IntegerArray(1, 5);
        int v = 0;
        for (int i = 0; i < owned.size(); i++) {
            owned.set(i, v += 2);
        }
        System.out.print("owned:");
        for (var x : owned) {
            System.out.print(" " + x);
        }
        System.out.println();
    }

    static void testIntegerWidths() {
        System.out.println("-- integer widths --");

        ByteArray bytes = TestPackage.makeBytes();
        System.out.print("bytes:");
        for (var x : bytes) {
            System.out.print(" " + x);
        }
        System.out.println();

        ShortArray shorts = TestPackage.makeShorts();
        System.out.print("shorts:");
        for (var x : shorts) {
            System.out.print(" " + x);
        }
        System.out.println();

        // The 64-bit values exceed 32 bits, proving the elements are not
        // truncated.
        LongArray longs = TestPackage.makeLongs();
        System.out.print("longs:");
        for (var x : longs) {
            System.out.print(" " + x);
        }
        System.out.println();
    }

    static void testFloatingPointArrays() {
        System.out.println("-- floating-point arrays --");

        FloatArray floats = TestPackage.makeFloats();
        System.out.print("floats:");
        for (var x : floats) {
            System.out.print(" " + x);
        }
        System.out.println();
        System.out.println("sum: " + TestPackage.sumFloats(floats));

        // An `in out` parameter mutates the shared buffer in place.
        TestPackage.scaleFloats(floats);
        System.out.print("scaled:");
        for (var x : floats) {
            System.out.print(" " + x);
        }
        System.out.println();
        System.out.println("sum: " + TestPackage.sumFloats(floats));

        // A float array allocated on the Java side.
        FloatArray owned = new FloatArray(1, 3);
        owned.setUnslided(1, 0.25f);
        owned.setUnslided(2, 0.5f);
        owned.setUnslided(3, 0.75f);
        System.out.print("owned:");
        for (var x : owned) {
            System.out.print(" " + x);
        }
        System.out.println();

        DoubleArray doubles = TestPackage.makeDoubles();
        System.out.print("doubles:");
        for (var x : doubles) {
            System.out.print(" " + x);
        }
        System.out.println();
        System.out.println("sum: " + TestPackage.sumDoubles(doubles));
    }

    static void testBooleanArrays() {
        System.out.println("-- boolean arrays --");

        BooleanArray bools = TestPackage.makeBools();
        System.out.print("bools:");
        for (var x : bools) {
            System.out.print(" " + x);
        }
        System.out.println();
        System.out.println("any: " + TestPackage.anyTrue(bools));

        // An `in out` parameter flips each element in place.
        TestPackage.negateBools(bools);
        System.out.print("negated:");
        for (var x : bools) {
            System.out.print(" " + x);
        }
        System.out.println();
        System.out.println("any: " + TestPackage.anyTrue(bools));

        // A boolean array allocated on the Java side.
        BooleanArray owned = new BooleanArray(1, 2);
        owned.setUnslided(1, true);
        owned.setUnslided(2, false);
        System.out.print("owned:");
        for (var x : owned) {
            System.out.print(" " + x);
        }
        System.out.println();
    }

    static void testRecordArrays() {
        System.out.println("-- record arrays --");

        MyInt.Array arr = TestPackage.myIntArrFunc();
        System.out.println("bounds: " + arr.getBegin() + " " + arr.getEnd());
        System.out.print("content:");
        for (var el : arr) {
            System.out.print(" " + el.getI());
        }
        System.out.println();

        // An `in out` parameter mutates the records in place.
        TestPackage.myIntArrProc(arr);
        System.out.print("tripled:");
        for (var el : arr) {
            System.out.print(" " + el.getI());
        }
        System.out.println();
        TestPackage.printImage(arr);

        // A record array allocated and filled on the Java side.
        MyInt.Array owned = new MyInt.Array(1, 3);
        for (int i = 0; i < owned.size(); i++) {
            owned.set(i, new MyInt((i + 1) * 10));
        }
        System.out.print("owned:");
        for (var el : owned) {
            System.out.print(" " + el.getI());
        }
        System.out.println();
    }

    public static void main(String[] args) throws Throwable {
        testIntegerArrays();
        testIntegerWidths();
        testFloatingPointArrays();
        testBooleanArrays();
        testRecordArrays();
    }
}
